File: 498.diagonal-traverse.go
package leetcode

import "fmt"

/*
 * @lc app=leetcode id=498 lang=golang
 *
 * [498] Diagonal Traverse
 */

// @lc code=start
type Direction int
const (
	Up Direction = iota 
	Down 
)

func findDiagonalOrder(mat [][]int) []int {
	m := len(mat); n := len(mat[0])
	result := make([]int, m*n)
	i := 0; j := 0; curDir := Up
	NextPosition := func(i, j int, curDir Direction) (int, int, Direction){
		switch curDir{
		case Up:
			if  j == n -1 {
				//At the right edge 
				return i+1, j, Down 
			} else if i == 0{
				//At the top edge
				return i, j+1, Down
			} else{
				//Anywhere else contine in the same direction
				return i - 1, j+1, Up
			}

		case Down:
			if i == m - 1{
				//At the Bottom edge
				return i, j+1, Up
			} else if j == 0 {
				//At the left edge
				return i+1, j, Up
			} else {
				//Continue on Down to the left 
				return i + 1, j - 1, Down
			}
		}
		panic("invalid direction")
	}
	for pos :=0 ; pos < m*n; pos++{
		result[pos] = mat[i][j]
		i, j, curDir = NextPosition(i, j, curDir)
	}

	return result
}
// @lc code=end

File: 498.diagonal-traverse.py
#
# @lc app=leetcode id=498 lang=python3
#
# [498] Diagonal Traverse
#

# @lc code=start
from typing import List

class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        if not mat or not mat[0]:
            return []
        
        m, n = len(mat), len(mat[0])
        result = []
        
        for d in range(m + n - 1):
            intermediate = []
            
            # Starting row and column for this diagonal
            r = 0 if d < n else d - n + 1
            c = d if d < n else n - 1
            
            # Collect all elements along this diagonal
            while r < m and c >= 0:
                intermediate.append(mat[r][c])
                r += 1
                c -= 1
            
            # Reverse on even diagonals
            if d % 2 == 0:
                result.extend(intermediate[::-1])
            else:
                result.extend(intermediate)
        
        return result

        
# @lc code=end

